contrib/babelfishpg_tsql/runtime/functions.c

```diff
@@ -2088,6 +2089,292 @@ numeric_radians(PG_FUNCTION_ARGS)
 	PG_RETURN_NUMERIC(result);
 }
 
+static inline int32_t
+GetUTF8CodePoint(const unsigned char *in, int len, int *consumed_p)
+{
+	int32_t		code;
+	int		consumed;
+	
+	if (len == 0)
+		return EOF;
+	
+	if ((in[0] & 0x80) == 0)
+	{
+		/* 1 byte - 0xxxxxxx */
+		code = in[0];
+		consumed = 1;
+	}
+	else if ((in[0] & 0xE0) == 0xC0)
+	{
+		/* 2 byte - 110xxxxx 10xxxxxx */
+		if (len < 2)
+			ereport(ERROR,
+					(errcode(ERRCODE_DATA_EXCEPTION),
+					 errmsg("truncated UTF8 byte sequence starting with 0x%02x",
+							in[0])));
+		if ((in[1] & 0xC0) != 0x80)
+			ereport(ERROR,
+					(errcode(ERRCODE_DATA_EXCEPTION),
+					 errmsg("invalid UTF8 byte sequence starting with 0x%02x",
+							in[0])));
+		code = ((in[0] & 0x1F) << 6) | (in[1] & 0x3F);
+		consumed = 2;
+	}
+	else if ((in[0] & 0xF0) == 0xE0)
+	{
+		/* 3 byte - 1110xxxx 10xxxxxx 10xxxxxx */
+		if (len < 3)
+			ereport(ERROR,
+					(errcode(ERRCODE_DATA_EXCEPTION),
+					 errmsg("truncated UTF8 byte sequence starting with 0x%02x",
+							in[0])));
+		if ((in[1] & 0xC0) != 0x80 || (in[2] & 0xC0) != 0x80)
+			ereport(ERROR,
+					(errcode(ERRCODE_DATA_EXCEPTION),
+					 errmsg("invalid UTF8 byte sequence starting with 0x%02x",
+							in[0])));
+		code = ((in[0] & 0x0F) << 12) | ((in[1] & 0x3F) << 6) | (in[2] & 0x3F);
+		consumed = 3;
+	}
+	else if ((in[0] & 0xF8) == 0xF0)
+	{
+		/* 4 byte - 1110xxxx 10xxxxxx 10xxxxxx 10xxxxxx */
+		if (len < 4)
+			ereport(ERROR,
+					(errcode(ERRCODE_DATA_EXCEPTION),
+					 errmsg("truncated UTF8 byte sequence starting with 0x%02x",
+							in[0])));
+		if ((in[1] & 0xC0) != 0x80 || (in[2] & 0xC0) != 0x80 ||
+			(in[3] & 0xC0) != 0x80)
+			ereport(ERROR,
+					(errcode(ERRCODE_DATA_EXCEPTION),
+					 errmsg("invalid UTF8 byte sequence starting with 0x%02x",
+							in[0])));
+		code = ((in[0] & 0x07) << 18) | ((in[1] & 0x3F) << 12) |
```

Comment:
for those 0x07, and 18, a lot of bit operations, I can't figure out the real meaning when reading code
Is it possible to give some meaningful define ?